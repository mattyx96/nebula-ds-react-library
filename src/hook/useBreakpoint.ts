import {useState, useEffect} from 'react';


const getCurrentBreakpoints = () => {
  const width = window.innerWidth;

  if (width < 640) return 'xs';
  if (width >= 640 && width < 768) return 'sm';
  if (width >= 768 && width < 1024) return 'md';
  if (width >= 1024 && width < 1280) return 'lg';
  if (width >= 1280 && width < 1536) return 'xl';
  return '2xl';
};

export const useBreakpoint = () => {
  const [breakpoint, setBreakpoint] = useState(getCurrentBreakpoints());

  useEffect(() => {
    const resizeListener = () => {
      setBreakpoint(getCurrentBreakpoints());
    };

    window.addEventListener('resize', resizeListener);
    return () => {
      window.removeEventListener('resize', resizeListener);
    };
  }, []);

  return {
    current: breakpoint,
    isDesktop: ['lg', 'xl', 'xxl', 'xxxl'].includes(breakpoint),
    isMobile: ['sm', 'xs', 'xxs'].includes(breakpoint),
    isTablet: breakpoint === 'md',
  };
};
