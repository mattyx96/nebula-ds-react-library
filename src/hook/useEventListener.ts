import {useEffect} from 'react'

export const useEventListener = (event: string, listener: () => void, useCapture?: boolean) => {
  useEffect(() => {
    if (listener) {
      listener()
      window.addEventListener(event, listener, useCapture)

      return () => window.removeEventListener(event, listener, useCapture)
    }

    return () => {
    }
  }, [event, listener, useCapture])
}
