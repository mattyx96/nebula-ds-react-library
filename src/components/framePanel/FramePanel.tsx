import {ReactNode} from "react";
import {FrameConnector, FrameConnectorNode} from "../frameConnector/FrameConnector";
import {Text} from "../typography/Typography";
import {useBreakpoint} from "../../hook/useBreakpoint.ts";
import useElementDimensions from "../../hook/useElementDimentions.ts";
import {lightJsTokens} from "nebula-ds-tokens";

type Props = ({ title: string; renderTitle?: undefined } | { renderTitle: ReactNode; title?: undefined }) & {
  className?: string
  children: ReactNode
  renderHeader: ReactNode
  renderFooter?: ReactNode
  renderSide?: ReactNode
  inverse?: boolean
  //add renderFooter
  //add renderSide
}

export const FramePanel = (props: Props) => {
  const breakpoint = useBreakpoint();
  const {dimensions, ref} = useElementDimensions();

  const frameConnectorSize = breakpoint.isDesktop ? 'M' : 'S';
  const isHeaderFrameConnectorOneNodeWidth = () => {
    if (breakpoint.isDesktop) {
      return dimensions.width <= 130
    } else {
      return dimensions.width <= 90
    }
  }

  const isHeaderFrameConnectorHidden = () => {
    if (breakpoint.isDesktop) {
      return dimensions.width <= 70
    } else {
      return dimensions.width <= 50
    }
  }

  return (
    <div className={`w-full flex flex-col ${props.className}`}>

      {/* header */}
      <div className="flex w-full gap-4 justify-between items-center">
        {props.inverse && (
          <>
            <div className="flex gap-4 items-center">
              {props.renderHeader}
            </div>
            <div ref={ref} className="w-full flex flex-1 scale-[-1]">
              {!isHeaderFrameConnectorHidden() &&
                <FrameConnector
                  size={frameConnectorSize}
                  firstNode={{hidden: isHeaderFrameConnectorOneNodeWidth()}}
                />
              }
            </div>
          </>)}

        {props.renderTitle
          ? props.renderTitle
          : <Text component="h1" variant="header1"
                  className="!text-2xl md:!text-3xl xl:!text-5xl !leading-0">{props.title}</Text>
        }

        {!props.inverse && (
          <>
            <div ref={ref} className="w-full flex flex-1 scale-[-1]">
              {!isHeaderFrameConnectorHidden() &&
                <FrameConnector
                  size={frameConnectorSize}
                  secondNode={{hidden: isHeaderFrameConnectorOneNodeWidth()}}
                />
              }
            </div>
            <div className="flex gap-4 items-center">
              {props.renderHeader}
            </div>
          </>)}
      </div>
      {/* end header */}

      {/* body container */}
      <div className={`flex flex-1 flex-col`}>

        {/* main vertical container (used for mobile side actions )*/}
        <div className={`flex flex-1 flex-col ${props.inverse ? 'items-end' : 'items-start'}`}>

          <div className={`flex flex-1 gap-4 mb-4`}>
            {breakpoint.isMobile && props.renderSide}
          </div>

          {
            breakpoint.isMobile && (
              <>
                {
                  props.inverse
                    ? <FrameConnectorNode
                      size={frameConnectorSize}
                      className="!inline !rotate-[270deg] !origin-center !h-[44px]"
                      fill={lightJsTokens.nbFrameBackgroundPrimary}
                    />
                    : <FrameConnectorNode
                      size={frameConnectorSize}
                      className="!inline !w-fit !scale-x-[-1] !rotate-90 !origin-center"
                      fill={lightJsTokens.nbFrameBackgroundPrimary}
                    />
                }
              </>
            )
          }
        </div>
        {/*end main vertical container (used for mobile side actions )*/}

        {/* level 2 - body here */}
        <div className=" flex h-full items-stretch">
          {/* non-inverse side actions and connector */}
          {!props.inverse &&
            <div className="flex flex-col justify-items-stretch gap-4">
              {!breakpoint.isMobile && props.renderSide}
              <FrameConnector
                size={frameConnectorSize}
                className="flex flex-1 items-stretch"
                divider={breakpoint.isMobile}
                vertical
              />
            </div>
          }

          {/* body */}
          <div className=" flex flex-1 flex-col gap-4 pt-4">
            <div className=" flex flex-1 px-4">
              {props.children}
            </div>
            {/* footer */}
            <div className="flex items-end gap-4 h-fit">
              {props.inverse && props.renderFooter}
              <FrameConnector
                className="h-fit"
                size={frameConnectorSize}
                {...(props.inverse
                    ? {
                      secondNode: {hidden: true},
                      firstNode: {hidden: !props.renderFooter}
                    }
                    : {
                      firstNode: {hidden: true},
                      secondNode: {hidden: !props.renderFooter}
                    }
                )}
              />
              {!props.inverse && props.renderFooter}
            </div>
            {/* end footer */}
          </div>
          {/* end body */}

          {/* inverse side actions and connector */}
          {props.inverse &&
            <div className="flex flex-col justify-items-stretch gap-4">
              {!breakpoint.isMobile && props.renderSide}
              <FrameConnector
                size={frameConnectorSize}
                className="flex flex-1 items-stretch scale-[-1]"
                divider={breakpoint.isMobile}
                vertical
              />
            </div>
          }
        </div>
        {/* end level 2 - body here */}

      </div>
      {/* end body container */}

    </div>
  )
}
