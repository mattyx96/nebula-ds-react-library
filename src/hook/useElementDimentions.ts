import {useEventListener} from './useEventListener'
import {useCallback, useRef, useState} from 'react'

type ElementDimensions = {
  bottom: number
  height: number
  left: number
  right: number
  top: number
  width: number
  x: number
  y: number
}

const useElementDimensions = () => {
  const ref = useRef(null)
  const [dimensions, setDimensions] = useState<ElementDimensions>({
    bottom: 0,
    height: 0,
    left: 0,
    right: 0,
    top: 0,
    width: 0,
    x: 0,
    y: 0
  })

  const refresh = useCallback(() => {
    const domRect = (ref.current as unknown as Element)?.getBoundingClientRect()

    if (domRect) {
      setDimensions(domRect)
    }
  }, [])

  useEventListener('resize', refresh);
  useEventListener('scroll', refresh, true);

  return {dimensions, ref, refresh}
}

export default useElementDimensions
