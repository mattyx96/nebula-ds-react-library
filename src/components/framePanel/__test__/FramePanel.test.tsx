import {render, screen} from '@testing-library/react';
import {describe, it, expect, vi, beforeEach, Mock} from 'vitest';
import {useBreakpoint} from '../../../hook/useBreakpoint';
import {FramePanel} from "../FramePanel.tsx";

describe('FramePanel with mocked useBreakpoint', () => {
  beforeEach(() => {
    (useBreakpoint as Mock).mockReturnValue(() => ({
      current: 'sm',
      isDesktop: false,
      isMobile: false,
      isTablet: false,
    }));
  });

  it('renders correctly on desktop', () => {
    const {container} = render(
      <FramePanel
        title="Desktop Test"
        renderHeader={<div>Header</div>}
        renderFooter={<div>Footer</div>}
      >
        <div className="h-96 bg-background-contrast-primary-500 w-full rounded-lg"/>
      </FramePanel>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Desktop Test')).toBeInTheDocument();
  });

  it('renders correctly on mobile', () => {
    // Simulate mobile view
    vi.mocked(useBreakpoint).mockReturnValue({
      current: 'sm',
      isDesktop: false,
      isMobile: true,
      isTablet: false,
    });

    const {container} = render(
      <FramePanel
        title="Mobile Test"
        renderHeader={<div>Header</div>}
        renderFooter={<div>Footer</div>}
      >
        <div className="h-96 bg-background-contrast-primary-500 w-full rounded-lg"/>
      </FramePanel>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Mobile Test')).toBeInTheDocument();
  });

  it('renders correctly on tablet', () => {
    // Simulate tablet view
    vi.mocked(useBreakpoint).mockReturnValue({
      current: 'md',
      isDesktop: false,
      isMobile: false,
      isTablet: true,
    });
    const {container} = render(
      <FramePanel
        title="Tablet Test"
        renderHeader={<div>Header</div>}
        renderFooter={<div>Footer</div>}
      >
        <div className="h-96 bg-background-contrast-primary-500 w-full rounded-lg"/>
      </FramePanel>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Tablet Test')).toBeInTheDocument();
  });
});
