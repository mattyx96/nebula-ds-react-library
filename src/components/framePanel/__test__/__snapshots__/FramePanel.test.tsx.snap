// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FramePanel with mocked useBreakpoint > renders correctly on desktop 1`] = `
<div>
  <div
    class="w-full flex gap-4 flex-col "
  >
    <div
      class="flex w-full gap-4 justify-between items-center "
    >
      <h1
        class="header1 !text-2xl md:!text-3xl xl:!text-5xl !leading-0"
        variant="header1"
      >
        Desktop Test
      </h1>
      <div
        class="w-full flex flex-1 scale-[-1]"
      />
      <div
        class="flex gap-4 items-center"
      >
        <div>
          Header
        </div>
      </div>
    </div>
    <div
      class="flex flex-1 flex-col "
    >
      <div
        class="flex h-full flex-grow items-stretch "
      >
        <div
          class="flex flex-col justify-items-stretch gap-4"
        >
          <div
            class="flex flex-col w-fit flex flex-1 items-stretch"
          >
            <div
              class="h-full flex-1 w-[4px] bg-frame-background-primary undefined"
            />
            <div
              class="flex items-center justify-center rotate-90 origin-center undefined"
            >
              <svg
                fill="none"
                viewBox="0 0 44 43"
                width="44"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M43.5 1.63276C43.5 0.731013 42.769 0 41.8672 0C30.896 0 22.002 8.89396 22.002 19.8652V21.7336C22.002 22.0226 21.9614 22.3108 21.8742 22.5863C19.2457 30.8903 12.2732 37.2627 3.62405 39.0358C3.43745 39.0741 3.2468 39.0918 3.05632 39.0918H0.599918C0.544689 39.0918 0.49992 39.1366 0.49992 39.1918V42.9C0.49992 42.9553 0.544693 43 0.599918 43H35.358C35.3867 43 35.415 42.9907 35.4437 42.9907H40.5C42.1569 42.9907 43.5 41.6476 43.5 39.9907V1.63276ZM22.002 38.9958C22.002 39.0488 21.9591 39.0918 21.9061 39.0918H21.8997C21.8047 39.0918 21.7655 38.96 21.8406 38.9017C21.9048 38.8517 22.002 38.8962 22.002 38.9776V38.9958Z"
                  fill="#DEDED3"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="flex flex-1 flex-col flex-grow gap-4"
        >
          <div
            class="flex flex-1 px-4 "
          >
            <div
              class="h-96 bg-background-contrast-primary-500 w-full rounded-lg"
            />
          </div>
          <div
            class="flex items-end gap-4 h-fit "
          >
            <div
              class="w-full flex flex-1 h-fit"
            >
              <div
                class="flex items-end flex-1 h-fit"
              >
                <div
                  class="w-full h-[4px] bg-frame-background-primary undefined"
                />
              </div>
            </div>
            <div>
              Footer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FramePanel with mocked useBreakpoint > renders correctly on mobile 1`] = `
<div>
  <div
    class="w-full flex gap-4 flex-col "
  >
    <div
      class="flex w-full gap-4 justify-between items-center "
    >
      <h1
        class="header1 !text-2xl md:!text-3xl xl:!text-5xl !leading-0"
        variant="header1"
      >
        Mobile Test
      </h1>
      <div
        class="w-full flex flex-1 scale-[-1]"
      />
      <div
        class="flex gap-4 items-center"
      >
        <div>
          Header
        </div>
      </div>
    </div>
    <div
      class="flex flex-1 flex-col "
    >
      <div
        class="flex flex-col items-start"
      >
        <div
          class="flex items-center justify-center !inline !w-fit !scale-x-[-1] !rotate-90 !origin-center"
        >
          <svg
            fill="none"
            viewBox="0 0 44 43"
            width="44"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M43.5 1.63276C43.5 0.731013 42.769 0 41.8672 0C30.896 0 22.002 8.89396 22.002 19.8652V21.7336C22.002 22.0226 21.9614 22.3108 21.8742 22.5863C19.2457 30.8903 12.2732 37.2627 3.62405 39.0358C3.43745 39.0741 3.2468 39.0918 3.05632 39.0918H0.599918C0.544689 39.0918 0.49992 39.1366 0.49992 39.1918V42.9C0.49992 42.9553 0.544693 43 0.599918 43H35.358C35.3867 43 35.415 42.9907 35.4437 42.9907H40.5C42.1569 42.9907 43.5 41.6476 43.5 39.9907V1.63276ZM22.002 38.9958C22.002 39.0488 21.9591 39.0918 21.9061 39.0918H21.8997C21.8047 39.0918 21.7655 38.96 21.8406 38.9017C21.9048 38.8517 22.002 38.8962 22.002 38.9776V38.9958Z"
              fill="#DEDED3"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex h-full flex-grow items-stretch "
      >
        <div
          class="flex flex-col justify-items-stretch gap-4"
        >
          <div
            class="flex flex-col w-fit flex flex-1 items-stretch"
          >
            <div
              class="h-full flex-1 w-[4px] bg-frame-background-primary undefined"
            />
            <div
              class="flex items-center justify-center rotate-90 origin-center undefined"
            >
              <svg
                fill="none"
                viewBox="0 0 44 43"
                width="44"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M43.5 1.63276C43.5 0.731013 42.769 0 41.8672 0C30.896 0 22.002 8.89396 22.002 19.8652V21.7336C22.002 22.0226 21.9614 22.3108 21.8742 22.5863C19.2457 30.8903 12.2732 37.2627 3.62405 39.0358C3.43745 39.0741 3.2468 39.0918 3.05632 39.0918H0.599918C0.544689 39.0918 0.49992 39.1366 0.49992 39.1918V42.9C0.49992 42.9553 0.544693 43 0.599918 43H35.358C35.3867 43 35.415 42.9907 35.4437 42.9907H40.5C42.1569 42.9907 43.5 41.6476 43.5 39.9907V1.63276ZM22.002 38.9958C22.002 39.0488 21.9591 39.0918 21.9061 39.0918H21.8997C21.8047 39.0918 21.7655 38.96 21.8406 38.9017C21.9048 38.8517 22.002 38.8962 22.002 38.9776V38.9958Z"
                  fill="#DEDED3"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="flex flex-1 flex-col flex-grow gap-4"
        >
          <div
            class="flex flex-1 px-4 "
          >
            <div
              class="h-96 bg-background-contrast-primary-500 w-full rounded-lg"
            />
          </div>
          <div
            class="flex items-end gap-4 h-fit "
          >
            <div
              class="w-full flex flex-1 h-fit"
            >
              <div
                class="flex items-end flex-1 h-fit"
              >
                <div
                  class="w-full h-[4px] bg-frame-background-primary undefined"
                />
              </div>
            </div>
            <div>
              Footer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FramePanel with mocked useBreakpoint > renders correctly on tablet 1`] = `
<div>
  <div
    class="w-full flex gap-4 flex-col "
  >
    <div
      class="flex w-full gap-4 justify-between items-center "
    >
      <h1
        class="header1 !text-2xl md:!text-3xl xl:!text-5xl !leading-0"
        variant="header1"
      >
        Tablet Test
      </h1>
      <div
        class="w-full flex flex-1 scale-[-1]"
      />
      <div
        class="flex gap-4 items-center"
      >
        <div>
          Header
        </div>
      </div>
    </div>
    <div
      class="flex flex-1 flex-col "
    >
      <div
        class="flex h-full flex-grow items-stretch "
      >
        <div
          class="flex flex-col justify-items-stretch gap-4"
        >
          <div
            class="flex flex-col w-fit flex flex-1 items-stretch"
          >
            <div
              class="h-full flex-1 w-[4px] bg-frame-background-primary undefined"
            />
            <div
              class="flex items-center justify-center rotate-90 origin-center undefined"
            >
              <svg
                fill="none"
                viewBox="0 0 44 43"
                width="44"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M43.5 1.63276C43.5 0.731013 42.769 0 41.8672 0C30.896 0 22.002 8.89396 22.002 19.8652V21.7336C22.002 22.0226 21.9614 22.3108 21.8742 22.5863C19.2457 30.8903 12.2732 37.2627 3.62405 39.0358C3.43745 39.0741 3.2468 39.0918 3.05632 39.0918H0.599918C0.544689 39.0918 0.49992 39.1366 0.49992 39.1918V42.9C0.49992 42.9553 0.544693 43 0.599918 43H35.358C35.3867 43 35.415 42.9907 35.4437 42.9907H40.5C42.1569 42.9907 43.5 41.6476 43.5 39.9907V1.63276ZM22.002 38.9958C22.002 39.0488 21.9591 39.0918 21.9061 39.0918H21.8997C21.8047 39.0918 21.7655 38.96 21.8406 38.9017C21.9048 38.8517 22.002 38.8962 22.002 38.9776V38.9958Z"
                  fill="#DEDED3"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="flex flex-1 flex-col flex-grow gap-4"
        >
          <div
            class="flex flex-1 px-4 "
          >
            <div
              class="h-96 bg-background-contrast-primary-500 w-full rounded-lg"
            />
          </div>
          <div
            class="flex items-end gap-4 h-fit "
          >
            <div
              class="w-full flex flex-1 h-fit"
            >
              <div
                class="flex items-end flex-1 h-fit"
              >
                <div
                  class="w-full h-[4px] bg-frame-background-primary undefined"
                />
              </div>
            </div>
            <div>
              Footer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
